In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv("tips.csv")
data.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [3]:
data.shape

(244, 7)

# Observation:
- 244 customers have visited the restaurant and the data represent these 244 customers

## Which days the customers are visiting in high number?

In [4]:
# 1. Does the restaurant is working on all 7 days of the week?

data["day"].unique()

array(['Sun', 'Sat', 'Thur', 'Fri'], dtype=object)

In [5]:
## If i count Total sunday in the day column then it indicates that many customers have visited the restaurant

# Sun = 98. 98 customers have visited the restaurant on Sunday
# Sat = 
# Thur =
# Fri =

In [6]:
data["day"].value_counts()

Sat     87
Sun     76
Thur    62
Fri     19
Name: day, dtype: int64

## Observation:

- Saturday's have the highest visit to the restaurant by the customers

# Which time of the day is favorable among the customers?

In [7]:
# Lunch = count
# Dinner = count

In [8]:
data["time"].value_counts()

Dinner    176
Lunch      68
Name: time, dtype: int64

# People prefer to visit during Dinner as compared to lunch

## Whether Female prefer to visit the restaurant or Males prefer to this restaurant

In [10]:
data['sex'].value_counts()

Male      157
Female     87
Name: sex, dtype: int64

# Observation:
- The restaurant is popular among Male customers

## Filter the data to display all the Female smokers

In [11]:
data.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [12]:
data[(data['sex'] == "Female") & (data['smoker'] == "Yes")]

total_bill   tip     sex smoker   day    time  size
67         3.07  1.00  Female    Yes   Sat  Dinner     1
72        26.86  3.14  Female    Yes   Sat  Dinner     2
73        25.28  5.00  Female    Yes   Sat  Dinner     2
92         5.75  1.00  Female    Yes   Fri  Dinner     2
93        16.32  4.30  Female    Yes   Fri  Dinner     2
100       11.35  2.50  Female    Yes   Fri  Dinner     2
101       15.38  3.00  Female    Yes   Fri  Dinner     2
102       44.30  2.50  Female    Yes   Sat  Dinner     3
103       22.42  3.48  Female    Yes   Sat  Dinner     2
109       14.31  4.00  Female    Yes   Sat  Dinner     2
164       17.51  3.00  Female    Yes   Sun  Dinner     2
168       10.59  1.61  Female    Yes   Sat  Dinner     2
169       10.63  2.00  Female    Yes   Sat  Dinner     2
178        9.60  4.00  Female    Yes   Sun  Dinner     2
186       20.90  3.50  Female    Yes   Sun  Dinner     3
188       18.15  3.50  Female    Yes   Sun  Dinner     3
191       19.81  4.19  Female    Yes  Thur   Lunch     2
197       43.11  5.00  Female    Yes  Thur   Lunch     4
198       13.00  2.00  Female    Yes  Thur   Lunch     2
201       12.74  2.01  Female    Yes  Thur   Lunch     2
202       13.00  2.00  Female    Yes  Thur   Lunch     2
203       16.40  2.50  Female    Yes  Thur   Lunch     2
205       16.47  3.23  Female    Yes  Thur   Lunch     3
209       12.76  2.23  Female    Yes   Sat  Dinner     2
213       13.27  2.50  Female    Yes   Sat  Dinner     2
214       28.17  6.50  Female    Yes   Sat  Dinner     3
215       12.90  1.10  Female    Yes   Sat  Dinner     2
219       30.14  3.09  Female    Yes   Sat  Dinner     4
221       13.42  3.48  Female    Yes   Fri   Lunch     2
225       16.27  2.50  Female    Yes   Fri   Lunch     2
226       10.09  2.00  Female    Yes   Fri   Lunch     2
229       22.12  2.88  Female    Yes   Sat  Dinner     2
240       27.18  2.00  Female    Yes   Sat  Dinner     2

# Count the total number of female smoker visiting the restaurant?

In [ ]:
# data[condition][column name].funtion()

In [13]:
data[(data['sex'] == "Female") & (data['smoker'] == "Yes")]["sex"].count()

33

## Count the total number of smokers visiting the restaurant

In [14]:
data[data['smoker'] == "Yes"]['sex'].count()

93

## For each Gender, display the total bill

In [15]:
data.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [16]:
data[data["sex"] == "Male"]["total_bill"].sum()

3256.8199999999997

In [17]:
data[data["sex"] == "Female"]["total_bill"].sum()

1570.9499999999996

In [21]:
## Alternative way to get the same output as above

ser = data.groupby('sex')["total_bill"].sum()
ser

sex
Female    1570.95
Male      3256.82
Name: total_bill, dtype: float64

In [22]:
ser.astype('int')

sex
Female    1570
Male      3256
Name: total_bill, dtype: int32

## Display the total bills generated on each day of the week

In [23]:
data.groupby("day")["total_bill"].sum()

day
Fri      325.88
Sat     1778.40
Sun     1627.16
Thur    1096.33
Name: total_bill, dtype: float64

## Display the total and average bill generated on each day of the week

In [24]:
data.groupby("day")["total_bill"].agg(['sum', 'mean'])

sum       mean
day                     
Fri    325.88  17.151579
Sat   1778.40  20.441379
Sun   1627.16  21.410000
Thur  1096.33  17.682742

## For which day of the week, Males are generating the highest total bill and for which day of the week, Females are generating the highest total bill

In [25]:
data.groupby("sex")["total_bill"].max()

sex
Female    44.30
Male      50.81
Name: total_bill, dtype: float64

In [26]:
data.groupby(['day', 'sex'])['total_bill'].sum()

day   sex   
Fri   Female     127.31
      Male       198.57
Sat   Female     551.05
      Male      1227.35
Sun   Female     357.70
      Male      1269.46
Thur  Female     534.89
      Male       561.44
Name: total_bill, dtype: float64

In [27]:
data.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

## Display the count of customers for Lunch and Dinner on each day .

In [28]:
data.groupby(['sex','time','day']).count()

total_bill  tip  smoker  size
sex    time   day                                
Female Dinner Fri            5    5       5     5
              Sat           28   28      28    28
              Sun           18   18      18    18
              Thur           1    1       1     1
       Lunch  Fri            4    4       4     4
              Thur          31   31      31    31
Male   Dinner Fri            7    7       7     7
              Sat           59   59      59    59
              Sun           58   58      58    58
       Lunch  Fri            3    3       3     3
              Thur          30   30      30    30

In [29]:
data.groupby(['time','day'])["time"].count()

time    day 
Dinner  Fri     12
        Sat     87
        Sun     76
        Thur     1
Lunch   Fri      7
        Thur    61
Name: time, dtype: int64

In [31]:
data.groupby(["time", "day"])['sex'].count()

time    day 
Dinner  Fri     12
        Sat     87
        Sun     76
        Thur     1
Lunch   Fri      7
        Thur    61
Name: sex, dtype: int64

In [32]:
data.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

# Who gives more tip?
# Count of Male and Female at a particular time and particular day


In [34]:
data.groupby('smoker')['tip'].mean()

smoker
No     2.991854
Yes    3.008710
Name: tip, dtype: float64

In [35]:
data.groupby('smoker')['tip'].sum()

smoker
No     451.77
Yes    279.81
Name: tip, dtype: float64

In [36]:
data.groupby(['sex', 'time', 'day'])['sex'].count()

sex     time    day 
Female  Dinner  Fri      5
                Sat     28
                Sun     18
                Thur     1
        Lunch   Fri      4
                Thur    31
Male    Dinner  Fri      7
                Sat     59
                Sun     58
        Lunch   Fri      3
                Thur    30
Name: sex, dtype: int64

In [37]:
data.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [38]:
data['size'].value_counts()

2    156
3     38
4     37
5      5
1      4
6      4
Name: size, dtype: int64

# Display the percentage of customers who are smoker?

In [39]:
data['smoker'].value_counts(normalize = True) * 100

No     61.885246
Yes    38.114754
Name: smoker, dtype: float64

In [40]:
data.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

# What is the percentage of Male smokers?

In [42]:
data[data['smoker']=='Yes'].sex.value_counts(normalize=True) * 100

Male      64.516129
Female    35.483871
Name: sex, dtype: float64

In [44]:
data.groupby('smoker').sex.value_counts(normalize=True) * 100

smoker  sex   
No      Male      64.238411
        Female    35.761589
Yes     Male      64.516129
        Female    35.483871
Name: sex, dtype: float64

In [45]:
# 3 mins using group by 

In [59]:
data.groupby('sex')['smoker'].count() / len(data) * 100

sex
Female    35.655738
Male      64.344262
Name: smoker, dtype: float64

In [58]:
data.groupby('sex')['smoker'].count()["Male"] /len(data) * 100

64.34426229508196

In [54]:
ser = pd.Series([100, 200], index = ["M", "F"])
ser

M    100
F    200
dtype: int64

In [56]:
ser['M']

100